lib/pos_blockchain.go

```diff
@@ -408,20 +406,50 @@ func (bc *Blockchain) addBlockToBlockIndex(desoBlock *MsgDeSoBlock) error {
 }
 
 // shouldReorg determines if we should reorg to the block provided. We should reorg if
-// this block has a higher QC than our current tip and extends from either the committed
-// tip OR any uncommitted safe block in our block index.
-func (bc *Blockchain) shouldReorg(desoBlock *MsgDeSoBlock) bool {
-	return false
+// this block is proposed in a view greater than or equal to the currentView. Other
+// functions have validated that this block is not extending from a committed block
+// that is not the latest committed block, so there is no need to validate that here.
+func (bc *Blockchain) shouldReorg(desoBlock *MsgDeSoBlock, currentView uint64) bool {
+	chainTip := bc.GetBestChainTip()
+	// If this block extends from the chain tip, there's no need to reorg.
+	if chainTip.Hash.IsEqual(desoBlock.Header.PrevBlockHash) {
+		return false
+	}
+	// If the block is proposed in a view less than the current view, there's no need to reorg.
+	return desoBlock.Header.ProposedInView >= currentView
 }
 
-// handleReorg handles a reorg to the block provided. It does not check whether or not we should
+// tryReorgToNewTip handles a reorg to the block provided. It does not check whether or not we should
 // perform a reorg, so this should be called after shouldReorg. It will do the following:
-// 1. Update the bestChain and bestChainMap by removing blocks that are not uncommitted ancestor of this block.
+// 1. Update the bestChain and bestChainMap by removing blocks that are uncommitted and are not ancestors of this block.
 // 2. Update the bestChain and bestChainMap by adding blocks that are uncommitted ancestors of this block.
 // Note: addBlockToBestChain will be called after this to handle adding THIS block to the best chain.
-func (bc *Blockchain) handleReorg(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me.
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) tryReorgToNewTip(desoBlock *MsgDeSoBlock, currentView uint64, ancestorsToCommittedTip []*BlockNode) (_hasReorg bool, _err error) {
+	// Check if we should perform a reorg here
+	if !bc.shouldReorg(desoBlock, currentView) {
+		return false, nil
+	}
+	// For simplicity, we remove all uncommitted blocks and then re-add them.
+	highestCommittedBlock, idx := bc.getHighestCommittedBlock()
+	if highestCommittedBlock == nil {
+		// This is an edge case we'll never hit in practice since all the PoW blocks
+		// are committed.
+		return false, errors.New("tryReorgToNewTip: No committed blocks found")
+	}
```

Comment:
We don't need this check anymore. This check was only needed when this function previously had a loop that traced lineage back to the committed tip.

Reply:
hmm we still need to get the index of the committed tip so we can easily delete all the uncommitted blocks that come after it from `bestChain` and `bestChainMap`